src/utils/reporting_workflow.js

```diff
@@ -0,0 +1,195 @@
+  //   Knowledge Artifact producers SHALL specify the duration
+  //   in the PlanDefinition.action.timing element.
+
+  // For executing actions which are dependent on other actions,
+  //   Knowledge Artifact producers SHALL specify the related action
+  //   in the PlanDefinition.action.relatedAction element.
+
+  // The PlanDefinition.action.relatedAction.relationship to be used
+  //   for MedMorph related actions SHALL be before-start.
+  //   This implies that the related action cannot start
+  //   until the parent action is executed.
+
+  // To specify related actions which are delayed by a time duration,
+  //   the PlanDefinition.action.relatedAction.offsetDuration element SHALL be used.
+
+  // we use graphlib: https://github.com/dagrejs/graphlib/wiki/API-Reference
+  // to build a dependency graph, then do a preorder traversal to turn it into a list
+
+  // note this is a lot of complexity, and is really only ever necessary if there are
+  // actions with more than one relatedAction.
+
+  // TODO: what if there are 2 relatedActions with different offsets?
+
+  const g = new graphlib.Graph();
+
+  for (const action of planDefinition.action) {
+    g.setNode(action.id);
+
+    if (action.relatedAction) {
+      for (const relatedAction of action.relatedAction) {
+        // spec requires relationship SHALL === 'before-start'
+        // so relatedAction is after currentAction
+
+        // directed edge from arg1 --> arg2
+        // TODO: track the offset on the edge label maybe?
+        g.setEdge(action.id, relatedAction.actionId);
+        // note that relatedAction.id doesn't already have to be in the graph
+        // it will be added if not
+      }
+    }
+  }
+
+  // An implementation of topological sorting.
+  // Given a Graph g this function returns an array of nodes
+  // such that for each edge u -> v, u appears before v in the array.
+  // https://github.com/dagrejs/graphlib/wiki/API-Reference#alg-topsort
+  // note the graph must not have cycles
+  const path = graphlib.alg.topsort(g);
+  return path;
+}
+
+async function executeWorkflow(context) {
+  db.upsert(COLLECTION, context, c => c.id === context.id);
+
+  const planDef = context.planDefinition;
+  const ig = findProfile(planDef);
+
+  while (context.currentActionSequenceStep < context.actionSequence.length) {
+    const currentActionId = context.actionSequence[context.currentActionSequenceStep];
+    const action = planDef.action.find(a => a.id === currentActionId);
+
+    const actionCode = action.code[0].coding[0].code;
+
+    const execute = getFunction(ig, actionCode);
+
+    // TODO: maintain context at each step for debugging?
+    // context = deepCopy(context);
+
+    // check the DB to get cancelToken
+    // TODO: is there a better way to do this?
+    const cancelToken = db.select(COLLECTION, c => c.id === context.id).cancelToken;
+    if (cancelToken) return;
+
+    await execute(context);
```

Comment:
is it necessary for this to be await?  If so, do the functions in execute need to be async as well, or will this cover our bases with functions that use promises?

Reply:
We need to pick some standard for async operations to make sure things happen in the right order. I figured async here is the easiest, because then the actions that actually need to do async stuff, like call out to a network resource, can just tag themselves as async as well and use await. Not every action needs to be tagged async, this should still work even if they aren't